presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SourceStatsCollector.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.Column;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ColumnMetadata;
+import com.facebook.presto.spi.Constraint;
+import com.facebook.presto.spi.TableHandle;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.spi.statistics.TableStatistics;
+import com.facebook.presto.sql.planner.plan.InternalPlanVisitor;
+import com.google.common.collect.ImmutableList;
+
+import java.util.AbstractMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class SourceStatsCollector
+{
+    private final Metadata metadata;
+    private final Session session;
+
+    public SourceStatsCollector(Metadata metadata, Session session)
+    {
+        this.metadata = metadata;
+        this.session = session;
+    }
+
+    public Double collectSourceStats(PlanNode root)
+    {
+        SourceTableStatsVisitor sourceTableStatsVisitor = new SourceTableStatsVisitor();
+        root.accept(sourceTableStatsVisitor, null);
+
+        List<TableStatistics> sourceStatistics = sourceTableStatsVisitor.getTableStatistics();
+
+        if (sourceTableStatsVisitor.isSourceDataMissing()) {
+            return -1.0;
+        }
+
+        Iterator<TableStatistics> tableStatisticsIterator = sourceStatistics.iterator();
+        Double totalSourceSize = 0.0;
+        while (tableStatisticsIterator.hasNext()) {
+            TableStatistics ts = tableStatisticsIterator.next();
+            totalSourceSize = totalSourceSize + ts.getTotalSize().getValue();
+        }
```

Comment:
Would love to have @rschlussel 's comment on this. Because hash partition count could be a general estimation going beyond presto on spark. Would it possible to add a rule to our optimizer to benefit the entire stack (for both presto and presto on spark) with all the stages as well as future HBO estimation?

Reply:
@rschlussel : it would be wonderful to have your thoughts. Additionally I would love to have a mechanism where we can integrate this with the HBO where the current estimation sets the initial values and the HBO optimizes over the historical dataset and we can optimize between the two.

Reply:
You should at https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsCalculator.java#L51 for integration with HBO. It combines the heuristic based statistics with history based. 

I agree that a generic rule for setting hash partition count would be nice.  It would have to be designed with a pluggable cost model, though, because the costs associated with hash partition counts are different between the two. In streaming Presto the cost of under partitioning are higher (latency from decreased parallelism is a bigger deal, higher memory is a bigger problem, worse fault tolerance if it causes flakiness).  Also, except for bucketed execution, Presto doesn't have a distinction between how many nodes are executing the query (executors for Spark), and how granular the partitioning is (hash partition count), which makes the calculus different.  And on the other side, streaming Presto doesn't have the executor start up costs that PoS has.  Given the risks for streaming Presto of reducing the partition count, I'm not sure the heuristics based approach would be worth it. But, once it's integrated with more reliable stats from HBO, it could be very valuable.  Especially if we set hash partition counts per stage.